File: implementations/ios-sdk/uitests/Tests/ScreenTrackingTests.swift
import XCTest

final class ScreenTrackingTests: XCTestCase {
    let app = XCUIApplication()

    override func setUp() {
        continueAfterFailure = false
        clearProfileState(app: app, requireFreshAppInstance: true)
    }

    /// Reads the `screen-event-log` element and returns its visible text
    /// (falling back to its accessibility label) as a string.
    private func getScreenEventLogText() -> String {
        let element = findElement("screen-event-log", app: app)
        XCTAssertTrue(element.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT),
                      "Element screen-event-log not found")
        return extractText(element)
    }

    func testTrackSingleViewVisit() {
        waitForElement(app.buttons["navigation-test-button"])
        app.buttons["navigation-test-button"].tap()

        waitForElement(app.buttons["go-to-view-one-button"])
        app.buttons["go-to-view-one-button"].tap()

        waitForElement(app.otherElements["navigation-view-test-one"])
        waitForElement(findElement("screen-event-log", app: app))

        waitForTextEquals("screen-event-log", expected: "NavigationHome,NavigationViewOne", app: app)
    }

    func testTrackMultipleViewVisitsInOrder() {
        waitForElement(app.buttons["navigation-test-button"])
        app.buttons["navigation-test-button"].tap()

        waitForElement(app.buttons["go-to-view-one-button"])
        app.buttons["go-to-view-one-button"].tap()

        waitForElement(app.otherElements["navigation-view-test-one"])
        waitForElement(app.buttons["go-to-view-two-button"])
        app.buttons["go-to-view-two-button"].tap()

        waitForElement(app.otherElements["navigation-view-test-two"])
        waitForElement(findElement("screen-event-log", app: app))

        let logText = waitForElementText("screen-event-log", app: app) { text in
            text.contains("NavigationViewTwo")
        }

        let viewOneIndex = logText.range(of: "NavigationViewOne").map {
            logText.distance(from: logText.startIndex, to: $0.lowerBound)
        } ?? -1
        let viewTwoIndex = logText.range(of: "NavigationViewTwo").map {
            logText.distance(from: logText.startIndex, to: $0.lowerBound)
        } ?? -1

        XCTAssertGreaterThanOrEqual(viewOneIndex, 0)
        XCTAssertGreaterThan(viewTwoIndex, viewOneIndex)
    }

    func testTrackRevisitingViewOneAfterViewTwo() {
        waitForElement(app.buttons["navigation-test-button"])
        app.buttons["navigation-test-button"].tap()

        waitForElement(app.buttons["go-to-view-one-button"])
        app.buttons["go-to-view-one-button"].tap()

        waitForElement(app.otherElements["navigation-view-test-one"])
        waitForElement(app.buttons["go-to-view-two-button"])
        app.buttons["go-to-view-two-button"].tap()

        waitForElement(app.otherElements["navigation-view-test-two"])

        // Trigger platform back navigation to return to view one
        app.navigationBars.buttons.element(boundBy: 0).tap()

        waitForElement(app.otherElements["navigation-view-test-one"])
        waitForElement(findElement("screen-event-log", app: app))

        waitForTextEquals(
            "screen-event-log",
            expected: "NavigationHome,NavigationViewOne,NavigationViewTwo,NavigationViewOne",
            app: app
        )
    }
}

File: implementations/android-sdk/uitests/src/main/kotlin/com/contentful/optimization/uitests/tests/ScreenTrackingTests.kt
package com.contentful.optimization.uitests.tests

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.uiautomator.By
import androidx.test.uiautomator.UiDevice
import com.contentful.optimization.uitests.support.TestHelpers
import com.contentful.optimization.uitests.support.clearProfileState
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ScreenTrackingTests {
    private lateinit var device: UiDevice

    @Before
    fun setUp() {
        device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
        clearProfileState(device, requireFreshAppInstance = true)
    }

    private fun navigateToTestScreen() {
        TestHelpers.waitAndTap(device, By.res("navigation-test-button"), TestHelpers.EXTENDED_TIMEOUT)
        TestHelpers.waitForElement(device, By.res("go-to-view-one-button"), TestHelpers.EXTENDED_TIMEOUT)
    }

    @Test
    fun testTrackSingleViewVisit() {
        navigateToTestScreen()
        TestHelpers.waitAndTap(device, By.res("go-to-view-one-button"))

        TestHelpers.waitForElement(device, By.res("navigation-view-test-one"), TestHelpers.EXTENDED_TIMEOUT)
        TestHelpers.waitForElement(device, By.res("screen-event-log"), TestHelpers.EXTENDED_TIMEOUT)
        TestHelpers.waitForTextEquals(
            device, "screen-event-log", "NavigationHome,NavigationViewOne",
            timeout = TestHelpers.EXTENDED_TIMEOUT,
        )
    }

    @Test
    fun testTrackMultipleViewVisitsInOrder() {
        navigateToTestScreen()
        TestHelpers.waitAndTap(device, By.res("go-to-view-one-button"))

        TestHelpers.waitForElement(device, By.res("navigation-view-test-one"), TestHelpers.EXTENDED_TIMEOUT)
        TestHelpers.waitAndTap(device, By.res("go-to-view-two-button"), TestHelpers.EXTENDED_TIMEOUT)

        TestHelpers.waitForElement(device, By.res("navigation-view-test-two"), TestHelpers.EXTENDED_TIMEOUT)
        TestHelpers.waitForElement(device, By.res("screen-event-log"), TestHelpers.EXTENDED_TIMEOUT)

        val logText = TestHelpers.waitForElementText(
            device, "screen-event-log", timeout = TestHelpers.EXTENDED_TIMEOUT,
        ) { text ->
            text.contains("NavigationViewTwo")
        }

        val viewOneIndex = logText.indexOf("NavigationViewOne")
        val viewTwoIndex = logText.indexOf("NavigationViewTwo")

        Assert.assertTrue("ViewOne not found in log", viewOneIndex >= 0)
        Assert.assertTrue("ViewTwo not found in log", viewTwoIndex >= 0)
        Assert.assertTrue("ViewOne should come before ViewTwo", viewOneIndex < viewTwoIndex)
    }

    @Test
    fun testTrackRevisitingViewOneAfterViewTwo() {
        navigateToTestScreen()
        TestHelpers.waitAndTap(device, By.res("go-to-view-one-button"))

        TestHelpers.waitForElement(device, By.res("navigation-view-test-one"), TestHelpers.EXTENDED_TIMEOUT)
        TestHelpers.waitAndTap(device, By.res("go-to-view-two-button"), TestHelpers.EXTENDED_TIMEOUT)

        TestHelpers.waitForElement(device, By.res("navigation-view-test-two"), TestHelpers.EXTENDED_TIMEOUT)

        device.pressBack()

        TestHelpers.waitForElement(device, By.res("navigation-view-test-one"), TestHelpers.EXTENDED_TIMEOUT)
        TestHelpers.waitForElement(device, By.res("screen-event-log"), TestHelpers.EXTENDED_TIMEOUT)
        TestHelpers.waitForTextEquals(
            device,
            "screen-event-log",
            "NavigationHome,NavigationViewOne,NavigationViewTwo,NavigationViewOne",
            timeout = TestHelpers.EXTENDED_TIMEOUT,
        )
    }
}
